File: src/com/zypex/piflow/profile/CompoundProfileSegment.java
package com.zypex.piflow.profile;

import utils.math.Vector;


import java.util.List;

public class CompoundProfileSegment extends ProfileSegment {

    List<ProfileSegment> segments;

    @Override
    public double upperBound() {
        double highest = segments.get(0).lowerBound();

        for(int i = 1; i < segments.size(); i++){
            final double current = segments.get(i).lowerBound();
            if(current > highest) highest = current;
        }

        return highest;
    }

    @Override
    public double lowerBound() {
        double lowest = segments.get(0).lowerBound();

        for(int i = 1; i < segments.size(); i++){
            final double current = segments.get(i).lowerBound();
            if(current < lowest) lowest = current;
        }

        return lowest;
    }

    @Override
    public double getLength() {
        double total = 0;
        for (ProfileSegment s : segments) total += s.getLength();
        return total;
    }

    @Override
    public CompoundProfileSegment offset(double offset) throws ClassCastException {
        CompoundProfileSegment out = new CompoundProfileSegment();
        for(ProfileSegment segment : segments){
            out.add(segment.offset(offset));
        }

        return out;
    }

    public void add(ProfileSegment segment) {
        if(segment instanceof CompoundProfileSegment){
            for(ProfileSegment sub : ((CompoundProfileSegment) segment).segments) add(sub);
        }

        segments.add(segment);
    }

    public void append(ProfileSegment segment){
        if(segment instanceof CompoundProfileSegment){
            for(ProfileSegment sub : ((CompoundProfileSegment) segment).segments) append(sub);
            return;
        }

        add(segment.offset(upperBound()));
    }

    @Override
    public double getT(Vector pos) {

        double t = segments.get(0).getT(pos);
        Vector closest = segments.get(0).get(t).position;

        for(int i = 1; i < segments.size(); i++){
            final ProfileSegment segment = segments.get(i);
            final double currentT = segment.getT(pos);
            final Vector currentPos = segment.get(currentT).position;

            if(currentPos.subtract(pos).getMagnitude() < closest.subtract(pos).getMagnitude()){
                t = currentT;
                closest = currentPos;
            }
        }

        return t;
    }

    @Override
    public Derivatives<Vector> get(double input) {
        return null;
    }
}

File: src/com/zypex/piflow/profile/CompoundProfileSegment.kt
package com.zypex.piflow.profile

import com.zypex.piflow.math.Vector


class CompoundProfileSegment : ProfileSegment() {
    var segments: MutableList<ProfileSegment> = ArrayList()
    override fun upperBound(): Double {
        var highest = segments[0].lowerBound()
        for (i in 1 until segments.size) {
            val current = segments[i].lowerBound()
            if (current > highest) highest = current
        }
        return highest
    }

    override fun lowerBound(): Double {
        var lowest = segments[0].lowerBound()
        for (i in 1 until segments.size) {
            val current = segments[i].lowerBound()
            if (current < lowest) lowest = current
        }
        return lowest
    }

    override val length: Double
        get() {
            var total = 0.0
            for (s in segments) total += s.length
            return total
        }

    @Throws(ClassCastException::class)
    override fun offset(offset: Double): CompoundProfileSegment {
        val out = CompoundProfileSegment()
        for (segment in segments) {
            out.add(segment.offset(offset))
        }
        return out
    }

    fun add(segment: ProfileSegment) {
        if (segment is CompoundProfileSegment) {
            for (sub in segment.segments) add(sub)
        }
        segments.add(segment)
    }

    fun append(segment: ProfileSegment) {
        if (segment is CompoundProfileSegment) {
            for (sub in segment.segments) append(sub)
            return
        }
        add(segment.offset(upperBound()))
    }

    override fun getT(pos: Vector): Double {
        var t = segments[0].getT(pos)
        var closest = segments[0](t).position
        for (i in 1 until segments.size) {
            val segment = segments[i]
            val currentT = segment.getT(pos)
            val currentPos = segment(currentT).position
            if (currentPos.subtract(pos).magnitude < closest.subtract(pos).magnitude) {
                t = currentT
                closest = currentPos
            }
        }
        return t
    }

    override fun invoke(input: Double): Derivatives<Vector> = TODO()
}